File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len = nums.length;
        int res[] = new int[len];

        int product = 1;
        res[0] = 1;
        for(int i = 1; i < len; i++){
            product *= nums[i-1];
            res[i] = product;
        }
        product = 1;
        for(int i = len-2; i >= 0; i--){
            product *= nums[i+1];
            res[i] *= product;
        }

        return res;
    }
}  

File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        """
                 []
            [1,]1     []1
    [1, 2,]2     [1,]2
[1,2,3]3[1, 2]3 [1,3]3 [1,]3
        """
        res = []

        def dfs(curr, exclude):
            if len(curr) == len(nums):
                res.append(curr[:])
                return

            for i in range(len(exclude)):
                if not exclude[i]:
                    curr.append(nums[i])
                    exclude[i] = True
                    dfs(curr, exclude)

                    exclude[i] = False
                    curr.pop()
        dfs([], [False] * (len(nums)))
        return res


